File: content/problems/0515-find-largest-value-in-each-tree-row/solutions/level-order-maximum/Solution.kt
class Solution {
    fun largestValues(root:TreeNode?):List<Int>{
        if(root==null)return emptyList();
        val answer=mutableListOf<Int>();
        val queue=ArrayDeque<TreeNode>();
        queue.addLast(root);
        while(queue.isNotEmpty()){
            var maximum=Int.MIN_VALUE;
            repeat(queue.size){
                val node=queue.removeFirst();
                maximum=maxOf(maximum,node.`val`);
                node.left?.let{
                    queue.addLast(it)
                };
                node.right?.let{
                    queue.addLast(it)
                }
            };
            answer.add(maximum)
        };
        return answer
    }
}

File: content/problems/0515-find-largest-value-in-each-tree-row/solutions/level-order-maximum/solution.cpp
class Solution {
    public: vector<int> largestValues(TreeNode* root){
        if(!root)return{
        };
        vector<int> answer;
        queue<TreeNode*> nodes;
        nodes.push(root);
        while(!nodes.empty()){
            int maximum=INT_MIN;
            for(int size=nodes.size();size>0;--size){
                TreeNode* node=nodes.front();
                nodes.pop();
                maximum=max(maximum,node->val);
                if(node->left)nodes.push(node->left);
                if(node->right)nodes.push(node->right);
            }
            answer.push_back(maximum);
        }
        return answer;
    }
};
